python/hyperon/exts/snet_io/snet_gate.py

```diff
@@ -137,9 +139,105 @@ def get_service_messages(self):
+        while i < len(input_str):
+            symbol = input_str[i]
+            add_symbol = symbol
+            num_of_symbols_in_line += 1
+            if symbol in str_symbols:
+                word_memory = ""
+                str_detected[symbol] = not str_detected[symbol]
+            elif symbol in special_symbols: # word ended so we need to check if this is function
+                if word_memory in list_of_functions:
+                    pp_func_res = pp_func(input_str[i:], num_of_tabs) # every function separately treated
+                    input_str = input_str.replace(input_str[i + 1:], pp_func_res) # though we have ran pp_func over current function's body it is not full so we need to loop through renewed str_input
+                    num_of_symbols_in_line = 0
+                word_memory = ""
+                if symbol == "\n":
+                    num_of_symbols_in_line = 0
+                    len_to_previous_break = 0
+                elif symbol == "\t":
+                    num_of_symbols_in_line -= 1
+            elif symbol == '(':
+                word_memory = ""
+                num_of_tabs += 1
+                # we will put a line break not for every '(', only if number of symbols in line exceeded threshold. PLus, to prevent several consequent line breaks, we're checking distance to previous line break
+                if (num_of_symbols_in_line > line_threshold) and (len_to_previous_break > (num_of_tabs + 2)):
+                    add_symbol = "\n" + "\t" * (num_of_tabs - 1) + add_symbol
+                    num_of_symbols_in_line = 0
+                    len_to_previous_break = 0
+            elif symbol == ')':
+                word_memory = ""
+                num_of_tabs -= 1
+                # if num_of_tabs == 0 then it should be the end of the current expression (for example, we have function's type and function definition in one string).
+                if (num_of_tabs == 0) and (len_to_previous_break > (num_of_tabs + 2)):
+                    add_symbol = add_symbol + "\n"
+                    num_of_symbols_in_line = 0
+                    len_to_previous_break = 0
+            else:
+                str_flag = False
+                for key in str_detected:
+                    str_flag = str_flag or str_detected[key]
+                # we are adding something to a memory only if this is not a part of string
+                word_memory += symbol * (not str_flag)
+            pretty_res += add_symbol
+            i += 1
+            len_to_previous_break += 1
+        pretty_res = pretty_res.replace("  ", " ")
+        pretty_res = pretty_res.replace("\t ", "\t")
+        pretty_res = pretty_res.replace("\n ", "\n")
+        return pretty_res
+
     def generate_callers_text(self):
-        # TODO: pretty print
-        return "\n".join([repr(e) for e in self.generate_callers()])
+        res = "\n".join([repr(e) for e in self.generate_callers()])
+        return [ValueAtom(res)]
```

Comment:
there is no need to wrap the result of this function into atom since it is supposed as a pure Python function, and it can be called via py-dot from metta

Reply:
Ok.